[visionOS] Safari YouTube audio location stays stationary when window is moved while viewing other tab (audio is not spatial)

mwyrzykowski opened the pull request:
#### 4c5bf811e466e891172125319aa6ef7fb4003795
<pre>
[visionOS] Safari YouTube audio location stays stationary when window is moved while viewing other tab (audio is not spatial)
<a href="https://bugs.webkit.org/show_bug.cgi?id=260364">https://bugs.webkit.org/show_bug.cgi?id=260364</a>
&lt;radar://112181086&gt;

Reviewed by Andy Estes.

Set the STSLabel when the viewer becomes invisible so audio follows the scene.

Set it back to nil when the viewer becomes visible to restore the current behavior.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setPageIsVisible):

* Source/WebCore/PAL/pal/spi/cocoa/AVFoundationSPI.h:
Stage SPI for non-internal clients.

Canonical link: <a href="https://commits.webkit.org/267018@main">https://commits.webkit.org/267018@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e81cf8fb3e1895715e1cd4398029c69cf6812bb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15391 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15695 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16058 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17146 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14440 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15538 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18212 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15793 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15575 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15991 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/13092 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17885 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13268 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13889 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/20830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14350 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14057 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17316 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14628 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/12398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13894 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3695 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18245 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14456 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -439,6 +445,16 @@ void getSupportedTypes(HashSet<String, ASCIICaseInsensitiveHash>& types) const f
+    if ([m_sampleBufferDisplayLayer respondsToSelector:@selector(setSTSLabel:)]) {
+        if (!visible) {
+            AVAudioSession *session = [PAL::getAVAudioSessionClass() sharedInstance];
+            [m_sampleBufferDisplayLayer setSTSLabel:session.spatialTrackingLabel];
```
This is behind `PLATFORM(VISION)` because `session.spatialTrackingLabel` is only available there

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -439,6 +445,16 @@ void getSupportedTypes(HashSet<String, ASCIICaseInsensitiveHash>& types) const f
         acceleratedRenderingStateChanged();
+
+#if PLATFORM(VISION)
+    if ([m_sampleBufferDisplayLayer respondsToSelector:@selector(setSTSLabel:)]) {
```
Nit: I'd leave a comment here noting that the `-respondsToSelector:` check can be removed after 114048644 is resolved.

mwyrzykowski replied:
Removed the respondsToSelector now that we are using SPI instead 👍

mwyrzykowski committed:
[visionOS] Safari YouTube audio location stays stationary when window is moved while viewing other tab (audio is not spatial)
https://bugs.webkit.org/show_bug.cgi?id=260364
<radar://112181086>

Reviewed by Andy Estes.

Set the STSLabel when the viewer becomes invisible so audio follows the scene.

Set it back to nil when the viewer becomes visible to restore the current behavior.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setPageIsVisible):

* Source/WebCore/PAL/pal/spi/cocoa/AVFoundationSPI.h:
Stage SPI for non-internal clients.

Canonical link: https://commits.webkit.org/267018@main
